lib/autoupdate/agent/installer.go

```diff
@@ -466,46 +526,261 @@ func (li *LocalInstaller) Link(ctx context.Context, version string) (revert func
+		}
+		if ok {
+			links = append(links, symlink{oldname, newname})
+		}
+		linked++
+	}
+	// bail if no binaries can be linked
+	if linked == 0 {
+		return trace.Errorf("no binaries available to link")
+	}
+
+	// validate that we can create the systemd service file before attempting linking
+
+	src := filepath.Join(svcDir, serviceName)
+	dst := filepath.Join(li.LinkServiceDir, serviceName)
+	err = tryCopy(dst, src, maxServiceFileSize)
+	if err != nil && !errors.Is(err, os.ErrExist) {
+		return trace.Errorf("error writing %s: %w", serviceName, err)
+	}
+
+	for _, link := range links {
+		if err := os.Symlink(link.oldname, link.newname); err != nil {
+			return trace.Errorf("failed to create symlink for %s: %w", filepath.Base(link.oldname), err)
+		}
+	}
+
+	return nil
+}
+
+// needsLink returns true when a symlink from oldname to newname needs to be created, or false if it exists.
+// If a non-symlink file or directory exists at newname, needsLink errors.
+// If a symlink to a different location exists, needsLink errors with ErrLinked.
+func needsLink(oldname, newname string) (ok bool, err error) {
+	exec, err := isExecutable(oldname)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if !exec {
+		return false, trace.Errorf("%s is not a regular executable file", oldname)
+	}
+	orig, err := os.Readlink(newname)
+	if errors.Is(err, os.ErrInvalid) ||
+		errors.Is(err, syscall.EINVAL) { // workaround missing ErrInvalid wrapper
+		// important: do not attempt to replace a non-linked install of Teleport
+		return false, trace.Errorf("refusing to replace file at %s", newname)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return true, nil
+	}
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if orig != oldname {
+		return false, trace.Errorf("refusing to replace link at %s: %w", newname, ErrLinked)
+	}
+	return false, nil
+}
+
+// tryCopy attempts to atomically copy a file from src to dst.
+// tryCopy errors if a file or directory already exists at dst.
+// tryCopy returns ErrLinked if a regular file smaller than n with different data already exists.
+// tryCopy returns os.ErrExist if a regular file smaller than n with the same contents exists.
+// The src and any dst file must be smaller than n, or tryCopy errors.
+func tryCopy(dst, src string, n int64) error {
+	srcData, err := readFileN(src, n)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fi, err := os.Lstat(dst)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return trace.Wrap(err)
+	}
+	if err == nil { // dst exists
+		if !fi.Mode().IsRegular() {
+			return trace.Errorf("refusing to replace irregular file at %s", dst)
+		}
+		dstData, err := readFileN(dst, n)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if bytes.Equal(srcData, dstData) {
+			return os.ErrExist
+		}
+		return trace.Errorf("refusing to replace file at %s: %w", dst, ErrLinked)
+	}
+	err = renameio.WriteFile(dst, srcData, 0644)
```

Comment:
There are a lot of hard-coded file mode values here. Would it make sense to create a few constants?